lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -1809,17 +1809,122 @@ bool SwiftLanguageRuntimeImpl::IsValidErrorValue(ValueObject &in_value) {
   return true;
 }
 
+static bool
+CheckIfAnyTypesDeclaredInLLDB(const swift::reflection::TypeRef *typeref) {
+  using swift::Demangle::Node;
+
+  const auto *protocol_typeref = llvm::dyn_cast_or_null<
+      const swift::reflection::ProtocolCompositionTypeRef>(typeref);
+
+  assert(protocol_typeref);
+  if (!protocol_typeref) {
+    assert(false);
+    return false;
+  }
+
+  swift::Demangle::Demangler dem;
+
+  NodePointer type = protocol_typeref->getDemangling(dem);
+  if (type->getKind() != Node::Kind::Type || type->getNumChildren() == 0)
+    return false;
+
+  NodePointer maybe_protocol_list = type->getFirstChild();
+  NodePointer protocol_list;
+
+  if (maybe_protocol_list->getKind() == Node::Kind::ProtocolList)
+    protocol_list = maybe_protocol_list;
+  else if ((maybe_protocol_list->getKind() ==
+                Node::Kind::ProtocolListWithAnyObject ||
+            maybe_protocol_list->getKind() ==
+                Node::Kind::ProtocolListWithClass) &&
+           maybe_protocol_list->getNumChildren() > 0)
+    protocol_list = maybe_protocol_list->getFirstChild();
+  else
+    return false;
+
+  if (protocol_list->getKind() != Node::Kind::ProtocolList ||
+      protocol_list->getNumChildren() == 0)
+    return false;
+
+  NodePointer type_list = protocol_list->getFirstChild();
+  if (type_list->getKind() != Node::Kind::TypeList)
+    return false;
+
+  for (size_t i = 0; i < type_list->getNumChildren(); ++i) {
+    NodePointer internal_type = type_list->getChild(i);
+    if (internal_type->getKind() != Node::Kind::Type ||
+        internal_type->getNumChildren() == 0)
+      continue;
+
+    NodePointer protocol = internal_type->getFirstChild();
+    if (protocol->getKind() != Node::Kind::Protocol ||
+        protocol->getNumChildren() == 0)
+      continue;
+
+    NodePointer module = protocol->getFirstChild();
+    if (module->getKind() != Node::Kind::Module)
+      continue;
+
+    if (module->getText().startswith("__lldb_expr_"))
```

Comment:
cf. https://github.com/apple/llvm-project/blob/03daa397f8ce5116b791b2ac932e4f8866a52654/lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp#L996

maybe this can be shared?

Reply:
to elaborate: clearly you can't share that function exactly because it uses AST types, but it looks like there's a case to add a virtual function to TypeSystemSwift. Feel free to do that in a follow-up commit though.